docs/design/exclusive-volume.md

```diff
@@ -0,0 +1,126 @@
+  - "@k82cn"
+  - "@zrss"
+approvers:
+  - "@k82cn"
+creation-date: 2019-11-21
+last-updated: 2019-11-21
+status: review
+---
+
+# Exclusive Volumes
+
+## Summary
+
+The purpose of this is to allow mount different volumes per pod of a Volcano Job.
+
+## Motivation
+
+Volume mount was supported from begin. But there are few limitations:
+
+1. We can specify volumes of Job scope, all the volumes are shared by all pods within a Job.
+
+2. We specify volumes by setting `TaskSpec.PodTemplateSpec.Volumes`, but similarly they are shared by pods within a task.
+
+But in real world, scenarios like DL/BigData, etc requires high performance. Shared storage has some performance issue, 
+like io limit, read/write conflicts. 
+
+Also some cloud vendors do not support volumes mounted to different nodes, it is to prevent data inconsistent.
+
+Given the above issues, we should support pods within volcano mounting volumes exclusively.
+
+## Goals
+
+Job TaskSpec will contain an additional structure called `VolumeClaimTemplates` to control exclusive volumes.
+
+## Proposal
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the TaskSpec for volcano Job.
+
+```go
+// TaskSpec specifies the task specification of Job
+type TaskSpec struct {
+    // Name specifies the name of tasks
+    Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+
+    // Replicas specifies the replicas of this TaskSpec in Job
+    Replicas int32 `json:"replicas,omitempty" protobuf:"bytes,2,opt,name=replicas"`
+
+    // Specifies the pod that will be created for this TaskSpec
+    // when executing a Job
+    Template v1.PodTemplateSpec `json:"template,omitempty" protobuf:"bytes,3,opt,name=template"`
+
+    // Specifies the lifecycle of task
+    // +optional
+    Policies []LifecyclePolicy `json:"policies,omitempty" protobuf:"bytes,4,opt,name=policies"`
+    
+    // volumeClaimTemplates is a list of claims that pods are allowed to reference.
+    // The Job controller is responsible for mapping network identities to
+    // claims in a way that maintains the identity of a pod. Every claim in
+    // this list must have at least one matching (by name) volumeMount in one
+    // container in the template. A claim in this list takes precedence over
+    // any volumes in the template, with the same name.
+    // TODO: Define the behavior if a claim already exists with the same name.
+    // +optional
+    VolumeClaimTemplates []v1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty" protobuf:"bytes,5,rep,name=volumeClaimTemplates"`
+}
+```
+
+- By default, this is empty. Will use that defined in `JobSpec` and `TaskSpec.Template`.
+
+- If `VolumeClaimTemplates` specified, these pvcs are referenced by all the pods of the task. But the persistent volume claim name is rebuilt by job controller. In a word, pods will not share any of the persistent volume claims. 
+
+- If the rebuilt persistent does not exist, job controller will create it.
```

Comment:
how about the lifecycle of pvc? will job controller delete it?

Reply:
Common as other children resource a job owns, i donot want it being different.   But the pv it binds can be retained by reclaim policy in  storage class.

Reply:
Meybe we should handle the pvc differently as it is stateful app.

Reply:
Hmm, it will look weird, inconsistent. as there are pvcs in job spec and pod spec, which are all deleted as volcano job deleted.

Reply:
@zrss Here we need your input